sql: sequences: add setval(name, value, is_called) builtin

vilterp opened the pull request:
Fixes #20410

Note that there are minor differences with PG because of #21564 

Release note: None

Commit:
sql: sequences: add setval(name, value, is_called) builtin

Fixes #20410

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21541)
<!-- Reviewable:end -->

justinj commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sequence.go, line 111 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21541#-L3BHKfSU2Jc71W1YXxq:-L3BHKfSU2Jc71W1YXxr:b-ggq7wt) ([raw file](https://github.com/cockroachdb/cockroach/blob/dca1452aafb6e756a4eceb588a560d2f7f967cd6/pkg/sql/sequence.go#L111)):*
> ```Go
> 	if !isCalled {
> 		newVal = newVal - descriptor.SequenceOpts.Increment
> 	}
> ```

Might be worth linking to that issue from the commit message here, since this has different behaviour from postgres if you set to `minvalue`?

---

*[pkg/sql/sem/builtins/builtins.go, line 1054 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21541#-L3BFUzRucHYYF5LoSvz:-L3BFUzSxLj_LUNsTJPu:bn3wdf) ([raw file](https://github.com/cockroachdb/cockroach/blob/dca1452aafb6e756a4eceb588a560d2f7f967cd6/pkg/sql/sem/builtins/builtins.go#L1054)):*
> ```Go
> 		},
> 		tree.Builtin{
> 			Types:      tree.ArgTypes{{"sequence_name", types.String}, {"value", types.Int}, {"is_caled", types.Bool}},
> ```

`s/is_caled/is_called/`

---

*[pkg/sql/sem/builtins/builtins.go, line 1071 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21541#-L3BFsAAaYwGXoFs15z9:-L3BFsAAaYwGXoFs15zA:b-1lqhb1) ([raw file](https://github.com/cockroachdb/cockroach/blob/dca1452aafb6e756a4eceb588a560d2f7f967cd6/pkg/sql/sem/builtins/builtins.go#L1071)):*
> ```Go
> 					return nil, err
> 				}
> 				return args[1], nil
> ```

Since we're calling this function in the logic test anyway, might as well include a test for this return value?

---

*[pkg/sql/sem/tree/eval.go, line 1995 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21541#-L3BKu8RrDuGhlwxloII:-L3BKu8RrDuGhlwxloIJ:b-jluisi) ([raw file](https://github.com/cockroachdb/cockroach/blob/dca1452aafb6e756a4eceb588a560d2f7f967cd6/pkg/sql/sem/tree/eval.go#L1995)):*
> ```Go
> 
> 	// SetSequenceValue sets the sequence's value.
> 	// if isCalled is false, the sequence is set such that the next time nextval() is called,
> ```

nit: `s/if/If/`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21541#-:-L3BLD9rAyPKA0hxU-SG:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sequence.go, line 111 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21541#-L3BHKfSU2Jc71W1YXxq:-L3Bd3sXnKrD02icxeAZ:bykh9kf) ([raw file](https://github.com/cockroachdb/cockroach/blob/dca1452aafb6e756a4eceb588a560d2f7f967cd6/pkg/sql/sequence.go#L111)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Might be worth linking to that issue from the commit message here, since this has different behaviour from postgres if you set to `minvalue`?
</blockquote></details>

Good point; linked to the issue from builtins.go. Also changed it so you can set the sequence to its minvalue, and added a couple tests.

---

*[pkg/sql/sem/builtins/builtins.go, line 1054 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21541#-L3BFUzRucHYYF5LoSvz:-L3B_XWhzG7LG2uJtTd8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dca1452aafb6e756a4eceb588a560d2f7f967cd6/pkg/sql/sem/builtins/builtins.go#L1054)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

`s/is_caled/is_called/`
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/eval.go, line 1995 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21541#-L3BKu8RrDuGhlwxloII:-L3B_e_rVn7J5HoXLd8n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dca1452aafb6e756a4eceb588a560d2f7f967cd6/pkg/sql/sem/tree/eval.go#L1995)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: `s/if/If/`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21541)*
<!-- Sent from Reviewable.io -->
